src/disk_store/storage.rs

```diff
@@ -0,0 +1,314 @@
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use serde::{Deserialize, Serialize};
+
+use super::file_writer::{BlobWriter, FileBlobWriter};
+use super::{ColumnLoader, PartitionMetadata, SubpartitionMetadata};
+use crate::logging_client::EventBuffer;
+use crate::mem_store::Column;
+use crate::perf_counter::{PerfCounter, QueryPerfCounter};
+
+#[derive(Serialize, Deserialize)]
+pub struct WALSegment<'a> {
+    pub id: u64,
+    pub data: Cow<'a, EventBuffer>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct MetaStore {
+    pub next_wal_id: u64,
+    pub partitions: Vec<PartitionMetadata>,
+}
+
+type PartitionID = u64;
+
+impl ColumnLoader for Storage {
+    fn load_column(
+        &self,
+        table_name: &str,
+        partition: PartitionID,
+        column_name: &str,
+        perf_counter: &QueryPerfCounter,
+    ) -> Vec<Column> {
+        Storage::load_column(self, partition, table_name, column_name, perf_counter)
+    }
+
+    fn load_column_range(
+        &self,
+        _start: PartitionID,
+        _end: PartitionID,
+        _column_name: &str,
+        _ldb: &crate::scheduler::InnerLocustDB,
+    ) {
+        todo!()
+    }
+}
+
+pub struct Storage {
+    wal_dir: PathBuf,
+    meta_db_path: PathBuf,
+    new_meta_db_path: PathBuf,
+    tables_path: PathBuf,
+    meta_store: Arc<Mutex<MetaStore>>,
+    writer: Box<dyn BlobWriter + Send + Sync + 'static>,
+    perf_counter: Arc<PerfCounter>,
+}
+
+impl Storage {
+    pub fn new(
+        path: &Path,
+        perf_counter: Arc<PerfCounter>,
+        readonly: bool,
+    ) -> (Storage, Vec<WALSegment>) {
+        let meta_db_path = path.join("meta");
+        let new_meta_db_path = path.join("meta_new");
+        let wal_dir = path.join("wal");
+        let tables_path = path.join("tables");
+        let writer = Box::new(FileBlobWriter::new());
+        let (meta_store, wal_segments) = Storage::recover(
+            &writer,
+            &meta_db_path,
+            &new_meta_db_path,
+            &wal_dir,
+            readonly,
+            perf_counter.as_ref(),
+        );
+        let meta_store = Arc::new(Mutex::new(meta_store));
+        (
+            Storage {
+                wal_dir,
+                meta_db_path,
+                new_meta_db_path,
+                tables_path,
+                meta_store,
+                writer,
+                perf_counter,
+            },
+            wal_segments,
+        )
+    }
+
+    fn recover<'a>(
+        writer: &FileBlobWriter,
+        mut meta_db_path: &'a Path,
+        new_meta_db_path: &'a Path,
+        wal_dir: &Path,
+        readonly: bool,
+        perf_counter: &PerfCounter,
+    ) -> (MetaStore, Vec<WALSegment<'static>>) {
+        // If db crashed during wal flush, might have new state at new_meta_db_path and potentially
+        // old state at meta_db_path. If both exist, delete old state and rename new state to old.
+        if writer.exists(new_meta_db_path).unwrap() {
```

Comment:
new meta db file might not be valid (partially written)
either guarantee atomic writes in `FileBlobWriter` or delete new meta db if old one still exists